sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -377,6 +377,7 @@ object DateTimeUtils {
     }
     val c = Calendar.getInstance()
     c.set(segments(0), segments(1) - 1, segments(2), 0, 0, 0)
-    Some((c.getTimeInMillis / 1000 / 3600 / 24).toInt)
+    c.set(Calendar.MILLISECOND, 0)
+    Some(millisToDays(c.getTimeInMillis))
```

Comment:
Instead of calling `millisToDays`, we could create a Calendar with UTC in the begging to avoid the timezone issues.